dkc/core/rest/folder.py

```diff
@@ -49,10 +58,93 @@ def perform_create(self, serializer: serializers.ModelSerializer):
         tree = parent.tree if parent else Tree.objects.create()
         serializer.save(tree=tree)
 
+    @swagger_auto_schema(
+        operation_description='Retrieve the path from the root folder to the requested folder.',
+        responses={200: FolderSerializer(many=True)},
+    )
     @action(detail=True)
     def path(self, request, pk=None):
         folder = self.get_object()
         # Start with the root folder
         ancestors = folder.ancestors[::-1]
         serializer = self.get_serializer(ancestors, many=True)
         return Response(serializer.data)
+
+    @swagger_auto_schema(
+        operation_description='Retrieve the terms of use for a folder.',
+        responses={
+            200: TermsSerializer,
+            204: 'No terms of use exist for the folder.',
+        },
+    )
+    @action(detail=True)
+    def terms(self, request, pk=None):
+        folder = self.get_object()
+        try:
+            terms = folder.tree.terms
+        except Terms.DoesNotExist:
+            return Response(status=204)
+
+        serializer = TermsSerializer(terms)
+        return Response(serializer.data)
+
+    @swagger_auto_schema(
+        operation_description='Retrieve terms of use if your user must agree to them.',
+        responses={
+            200: TermsSerializer,
+            204: 'No action is required; there are no terms, or you have already agreed.',
+        },
+    )
+    @action(detail=True, url_path='terms/agreement')
+    def terms_agreement(self, request, pk=None):
+        folder = self.get_object()
+        try:
+            terms = folder.tree.terms
+        except Terms.DoesNotExist:
+            return Response(status=204)  # No terms for the folder
+
+        # TODO get the actual user once REST authentication is in place
+        user = User.objects.first()
+
+        # TODO if no user is authenticated, just send out the terms.
+
+        try:
+            agreement = TermsAgreement.objects.get(terms=terms, user=user)
+            if agreement.checksum == terms.checksum:
+                return Response(status=204)  # User has already agreed
+        except TermsAgreement.DoesNotExist:
+            ...
+        serializer = TermsSerializer(terms)
+        return Response(serializer.data)
```

Comment:
Semantically, this should be a 404, since the agreement does not exist.

If the terms were inlined into the folder endpoint, there'd be no convenience of returning the term content here.

Reply:
I agree from a REST semantics standpoint, but dislike that in terms of HTTP semantics -- 4xx is supposed to indicate that the client did something wrong in their request. It shows up as errors in their front-end console, even though it's not an error, they're just essentially querying a boolean value. Is there some other way we could encode that boolean besides the HTTP status that you would find satisfactory?